### DISCUSSION TITLE
Why do we select only the 1st char in the decoded string in get_next_state() ?

### DISCUSSION THREAD

**Discussion by ekagra-ranjan** [outlines-dev/outlines#862] [Weird model output]
Hi! I faced an issue where in the FSM in CFG Guide is getting self.generation differently from what the model is outputting. Debugging it lead to [this line](https://github.com/outlines-dev/outlines/blob/a101c1c8066f7ce92cdbcbd46a269d06163eded9/outlines/fsm/guide.py#L412)

Shouldn;t it be just `self.generation += self.tokenizer.decode([token_id])` instead of `self.generation += self.tokenizer.decode([token_id])[0]` since the [decode function in HF](https://huggingface.co/docs/transformers/en/main_classes/tokenizer#transformers.PreTrainedTokenizer.decode) returns a str?

@rlouf

**Comment by ekagra-ranjan** [comment]
gentle reminder: @brandonwillard @rlouf

**Reply by brandonwillard** [reply]
@benlipkin and/or @lapp0 might know, but I don't off the top of my head.

**Comment by lapp0** [comment]
This appears to be a serious bug. Perhaps it's the root cause of https://github.com/outlines-dev/outlines/issues/796

It looks like the signature of `decode` was confused with the signature of `batch_decode`.

Will investigate as part of https://github.com/outlines-dev/outlines/issues/796

**Reply by lapp0** [reply]
@ekagra-ranjan I was mistaken in my initial review. This is an `outlines.models.tokenizer.Tokenizer` which returns `List[str]` for `decode()`

Could you please share a full reproduction script?
